src/query/api/v1/handler/prometheus/native/common.go

```diff
@@ -355,16 +308,41 @@ func RenderResultsJSON(
 			// would be at the result node but that would make it inefficient since
 			// we would need to create another block just for the sake of restricting
 			// the bounds.
-			if dp.Timestamp.Before(opts.Start) {
+			if dp.Timestamp.Before(opts.Start) || dp.Timestamp.After(opts.End) {
```

Comment:
One inconsistency w/ the prev implementation is that the filter-fn would not count data after `End` towards a series that would be included, but then in the normal rendering iteration it wouldn't skip these. Feels more correct to apply this in that path.